pkg/envoy/ads/response.go

```diff
@@ -18,74 +19,80 @@ const (
+		log.Error().Err(err).Msgf("[%s] Failed to create response for proxy with SerialNumber=%s on Pod with UID=%s", typeURI.Short(), proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+		xdsPathTimeTrack(startedAt, log.Debug(), typeURI, proxy, false)
 		return err
-	}
-
-	if err := (*server).Send(discoveryResponse); err != nil {
-		log.Error().Err(err).Msgf("[%s] Error sending to proxy with SerialNumber=%s on Pod with UID=%s", xdsShortName, proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+	} else if err := (*server).Send(discoveryResponse); err != nil {
+		log.Error().Err(err).Msgf("[%s] Error sending to proxy with SerialNumber=%s on Pod with UID=%s", typeURI.Short(), proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+		xdsPathTimeTrack(startedAt, log.Debug(), typeURI, proxy, false)
 		return err
 	}
 
-	success = true // read by deferred function
+	xdsPathTimeTrack(startedAt, log.Debug(), typeURI, proxy, true)
 	return nil
 }
 
 // sendResponse takes a set of TypeURIs which will be called to generate the xDS resources
 // for, and will have them sent to the proxy server.
 // If no DiscoveryRequest is passed, an empty one for the TypeURI is created
+// TODO(draychev): Convert to variadic function: https://github.com/openservicemesh/osm/issues/3127
 func (s *Server) sendResponse(typeURIsToSend mapset.Set,
 	proxy *envoy.Proxy,
 	server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer,
 	request *xds_discovery.DiscoveryRequest,
 	cfg configurator.Configurator) error {
-	success := true
-	if typeURIsToSend.Cardinality() == len(envoy.XDSResponseOrder) {
-		defer xdsPathTimeTrack(time.Now(), log.Info(), ADSUpdateStr, proxy.GetCertificateSerialNumber().String(), &success)
-	}
-
 	// Order is important: CDS, EDS, LDS, RDS
 	// See: https://github.com/envoyproxy/go-control-plane/issues/59
+	var typesToSend []envoy.TypeURI
 	for _, typeURI := range envoy.XDSResponseOrder {
-		if !typeURIsToSend.Contains(typeURI) {
-			continue
+		if typeURIsToSend.Contains(typeURI) {
+			typesToSend = append(typesToSend, typeURI)
 		}
+	}
+
+	if len(typesToSend) < 1 {
+		// This should never happen.
+		log.Error().Msg("The list of discovery responses passed to sendResponse() is empty. This is an invalid invocation of sendResponse().")
+		return errors.New("list of discovery responses is empty")
+	}
+
+	thereWereErrors := false
+
+	// Order is important: CDS, EDS, LDS, RDS
+	// See: https://github.com/envoyproxy/go-control-plane/issues/59
+	for _, typeURI := range typesToSend {
+		fullUpdateRequested := request != nil && envoy.TypeURI(request.TypeUrl).IsWildcard()
 
 		// Handle request when is not provided, and the SDS case
-		var finalReq *xds_discovery.DiscoveryRequest
-		if request == nil || request.TypeUrl == envoy.TypeWildcard.String() {
+		finalReq := request
+		if request == nil || fullUpdateRequested {
+			finalReq = &xds_discovery.DiscoveryRequest{TypeUrl: typeURI.String()}
 			if typeURI == envoy.TypeSDS {
```

Comment:
```
if  sds { 
   finalreq = makeRequestForAll...
} else {
  finalreq = &xds_discovery...
}
```
the first initialization gets overwritten is it's SDS type, o better skip allocating memory for the data structure (and probably will also avoid following pointer lifecycle since it is dereferenced).

Reply:
in fact the compiler probably already optimizes this, but wouldn't hurt to be mindful of these cases.

Reply:
In this PR I decided to eliminate the 2 ` } else {` statements below.
This, I believe, sacrifices CPU cycles for code readability.
I think it is a price worth paying (subjective) - especially when we have nested ifs (with elses).
I tend to optimize for humans reading the code and making improvements. We might be looking at this from different angles.

Reply:
I'm not sure this is much more readable. I guess I'm used to interpret the old logic and I'm biased?
before I interpreted it like:
```
if ConditionsToGenerateRequest {
  // Generate request
  if type == sds {
    // Special generate for SDS
  } else {
    // Generic generate request
  }
} else {
  // Use request past by param
}
```
To me this was more explicative and easy to follow. The only part is, this was not written to be understood this way by just reading it maybe.

Reply:
I'll revert this back to the version with the `} else {`.